tutorials/helper_functions.py

```diff
@@ -0,0 +1,222 @@
+    output folder, with one subfolder for each user, and subfolders
+    inside the user's folder for each data stream.
+    If a server failure happens, the function re-attempts the download.
+
+    Args:
+        keyring: a keyring generated by mano.keyring
+
+        users(iterable): A list of users to download data for.
+            If none are entered, it attempts to download data for all users
+
+        study_id(str): The id of a study
+
+        download_folder(str): path to a folder to download data
+
+        time_start(str): The initial date to download data
+            (Formatted in YYYY-MM-DD). Default is 2008-01-01, which is
+            before any Beiwe data existed.
+
+        time_end(str): The date to end downloads.
+            The default is today at midnight.
+
+        data_streams(iterable): A list of all data streams to download.
+            The default (None) is all possible data streams.
+
+    """
+    if study_id == "":
+        print("Error: Study ID is blank")
+        return
+
+    if (
+        keyring["USERNAME"] == ""
+        or keyring["PASSWORD"] == ""
+        or keyring["ACCESS_KEY"] == ""
+        or keyring["SECRET_KEY"] == ""
+    ):
+        print("Error: Did you set up the keyring_studies.py file?")
+        return
+
+    if not os.path.isdir(download_folder):
+        os.makedirs(download_folder, exist_ok=True)
+
+    if time_end is None:
+        time_end = datetime.today().strftime("%Y-%m-%d") + "T23:59:00"
+
+    if users is None:
+        print("Obtaining list of users...")
+        num_tries = 0
+        while num_tries < 5:
+            try:
+                users = [u for u in mano.users(keyring, study_id)]
+                break
+            except KeyboardInterrupt:
+                print("Someone closed the program")
+                sys.exit()
+            except mano.APIError as e:
+                print("Something is wrong with your credentials:")
+                print(e)
+                break
+            except Exception:
+                num_tries += 1
+
+    for u in users:
+        zf = None
+        num_tries = 0
+        while num_tries < 5:
+            print(f"Downloading data for {u}")
+            try:
+                zf = mano.sync.download(
+                    keyring,
+                    study_id,
+                    u,
+                    data_streams,
+                    time_start=time_start,
+                    time_end=time_end,
+                )
+                break
+            except requests.exceptions.ChunkedEncodingError:
+                print(f"Network failed in download of {u}, try {num_tries}")
+                num_tries += 1
+            except KeyboardInterrupt:
+                print("Someone closed the program")
+                sys.exit()
+            except mano.APIError as e:
+                print("Something is wrong with your credentials:")
+                print(e)
+                break
+
+        if num_tries == 5:
```

Comment:
I think while-else clause would be more readable
https://realpython.com/lessons/while-loop-else-clause/